src/PreparedStatementCache.cpp

```diff
@@ -0,0 +1,563 @@
+            LOG(Error) << error_desc.str() << endl;
+            Tango::Except::throw_exception(EXCEPTION_UNKNOWN_QUERY, error_desc.str(), __func__);
+        }
+    }
+
+    // to keep the compiler happy, we should never reach here
+    static string error = "Unknown";
+    return error;
+}
+
+//=============================================================================
+//=============================================================================
+string PreparedStatementCache::query_id_to_str(int data_type, int data_format, int data_write_type) const
+{
+    TRACE_LOGGER;
+    return create_query_id_from_data(data_type, data_format, data_write_type);
+}
+
+//=============================================================================
+//=============================================================================
+int PreparedStatementCache::get_insert_attr_ttl_bind_position(int data_write_type) const
+{
+    // this is ugly ugly ugly, but we can not bind the ttl by name in a
+    // statement that contains USING TLL ?. We are forced to bind by its
+    // position. To work around this for the InsertAttr_ queries, we simply
+    // calculate its point and feed it back to the main library.
+
+    // there are two scenarios (now we have condensed the statement), the
+    // InsertAttr_ type query has 10 bind points in the string, then an
+    // additional one or two based on data_write_type and finally the
+    // ttl bind point. This means the ttl is either 11 or 12 based on the
+    // data_write_type value. We can boil it down to a simple return below:
+
+    // 9 bind points before the data write type
+    int bind_point = 9;
+
+    if (data_write_type != Tango::WRITE)
+        bind_point++;
+
+    if (data_write_type != Tango::READ)
+        bind_point++;
+
+    // add one to get us to the ttl bind point
+    bind_point++;
+    return bind_point;
+}
+
+//=============================================================================
+//=============================================================================
+const string &PreparedStatementCache::look_up_query_string(Query query)
+{
+    TRACE_LOGGER;
+    const string &query_id = query_id_to_str(query);
+    auto query_iter = _query_cache.find(query_id);
+
+    // if we have not built the query, then we must first do this
+    // before it can be used for prepared statements by the user
+    if (query_iter == _query_cache.end())
+    {
+        ostringstream query_str;
+
+        switch (query)
+        {
+            case Query::GetAttrIdAndTtl:
+            {
+                query_str << "SELECT " << CONF_COL_ID << ", " << CONF_COL_TTL << " FROM "
+                          << _keyspace_name << "." << CONF_TABLE_NAME << " WHERE " << CONF_COL_NAME
+                          << " = ? AND " << CONF_COL_FACILITY << " = ?" << ends;
+            }
+            break;
+
+            case Query::GetAttrDataType:
+            {
+                // TODO remove ID/TTL
```

Comment:
Good idea.  Indeed id and ttl are never used when this statement is invoked.

Reply:
Yes, i wanted to call this out to you to double check. I will make the change as part of this pull request.